File: implementations/Y.java
// The types used are based on the following code: https://gist.github.com/aruld/3965968

public class Y{

    private static interface Func<T>{
	T apply (T x);
    }
    
    private static interface FuncToFunc<T>{
	Func<T> apply (FuncToFunc<T> x);
    }
    
    public static <T> Func<T> YComb (final Func<Func<T>> improver){ 
    	return ((FuncToFunc<T>) ((f) -> f.apply(f)))
    	    .apply((f) -> improver.apply((x) -> f.apply(f).apply(x)));
    };

    public static Func<Func<Integer>> facImprover = (partial) -> (x) -> (x == 1) ? 1 : x*partial.apply(x-1);
    public static Func<Func<Integer>> fibImprover = (partial) -> (x) -> (x < 2) ? 1 : partial.apply(x-1)+partial.apply(x-2);
    
    public static void main(String[] args) {
	Func<Integer> fac = YComb(facImprover);
	Func<Integer> fib = YComb(fibImprover);
	
	if(args.length != 1) return;
	int number = Integer.parseInt(args[0]);
	System.out.printf("fac(%d) = %d\n", number, fac.apply(number));
	System.out.printf("fib(%d) = %d\n", number, fib.apply(number));	
    }
}

File: implementations/y.erl
-module(y).
-export([main/0]).

%% @doc Y combinator implementation in Erlang.
%% To run this implementation, use the following command:
%%   erl -noshell -s y main -s init stop
%% @end

main() -> 
  Y = fun (F) ->
    (fun (X) -> 
      X(X) end)(fun (X) -> 
        F(fun (T) -> (X(X))(T) end) 
      end) 
    end,
  
  FacImprover = 
    fun(Partial) -> 
      fun 
          (N) when N < 2 -> 1; 
          (N) when N >= 2 -> N * Partial(N - 1) 
      end 
  end,

  FibImprover = 
    fun(Partial) -> 
      fun 
          (N) when N < 2 -> 1; 
          (N) when N >= 2 -> Partial(N - 2) + Partial(N - 1)
      end 
  end,

  Fac = Y(FacImprover),

  Fib = Y(FibImprover),

  io:format("Fac 0: ~p~n", [Fac(0)]),
  io:format("Fac 1: ~p~n", [Fac(1)]),
  io:format("Fac 2: ~p~n", [Fac(2)]),
  io:format("Fac 3: ~p~n", [Fac(3)]),
  io:format("Fac 4: ~p~n", [Fac(4)]),
  io:format("Fac 5: ~p~n", [Fac(5)]),

  io:format("Fib 0: ~p~n", [Fib(0)]),
  io:format("Fib 1: ~p~n", [Fib(1)]),
  io:format("Fib 2: ~p~n", [Fib(2)]),
  io:format("Fib 3: ~p~n", [Fib(3)]),
  io:format("Fib 4: ~p~n", [Fib(4)]),
  io:format("Fib 5: ~p~n", [Fib(5)]).

